x/slashing/src/keeper.rs

```diff
@@ -87,6 +103,211 @@ impl<SK: StoreKey, PSK: ParamsSubspaceKey, SSK: SlashingStakingKeeper<SK, M>, M:
         self.slashing_params_keeper.set(ctx, genesis.params);
     }
 
+    pub fn handle_validator_signature<DB: Database>(
+        &self,
+        ctx: &mut BlockContext<'_, DB, SK>,
+        addr_bytes: prost::bytes::Bytes,
+        power: u32,
+        signed: bool,
+    ) {
+        let height = ctx.height();
+
+        // fetch the validator public key
+        let cons_addr = ConsAddress::try_from(Vec::from(addr_bytes)).unwrap();
+        self.get_pub_key(ctx, &cons_addr)
+            .expect("validator consensus address not found");
+
+        // fetch signing info
+        let mut sign_info = self
+            .get_validator_signing_info(ctx, &cons_addr)
+            .expect("Expected signing info for validator but it is not found");
+
+        // this is a relative index, so it counts blocks the validator *should* have signed
+        // will use the 0-value default signing info if not present, except for start height
+        let params = self.slashing_params_keeper.get(ctx);
+        let index = sign_info.index_offset % params.signed_blocks_window as u32;
```

Comment:
Is it possible to make 'signed_blocks_window` a u32?

Reply:
I think that params struct should have types similar to cosmos and it is better to save type.

 I can change it to `u32` and then I have to change serialization/deserialization of params. 

What do you think about it?